FR #976: Add support for contract commitments (#1350)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/contractapplied.md

Author: In Sep 2 TF1, we discussed the name for this column.  `ContractApplied` didn't feel right to @flanakin and @ijurica, and we do have the precedent for using the suffix `Details` for JSON structures.  Thus, I am consider changing this column to `ContractDetails`.  

Regardless, that does leave the question of how to name the constituent object entries.  Consider `ContractCommitmentAppliedCost`.  We could go with `ContractCommitmentCost`, but that doesn't feel correct to me: it's the application of the contract to the Cost and Usage (either Purchase or Usage) that's relevant, so I would like something like that to be in the name.

Author: No further alternatives have been expressed in the past week for renaming the column.  We can certainly discuss this further in the coming weeks, but for now, I'm resolving this comment.

# Contract Applied

> Contract Applied is a set of four datapoints that associate a charge to one or more [*contract commitments*](#glossary:contract-commitment), denoted as key-value pairs in JSON format.  Contract Applied allows the practitioner to track the progress of the commitments to which they have agreed with a provider.

Reviewer: @shawnalpay a couple of questions:

Is there any sample data available in addition to the example shown in the PR?

I'm specifically interested in understanding what a Burn-Down Commitment Purchase record looks like. More precisely, I'm trying to figure out where we capture the ID, the amount of the commitment being acquired through that Purchase record, etc.

From what I can tell, the Contract Applied column currently seems to be intended only for Usage — or more precisely, for Contribution (since Purchase records can also contribute to a Build-Up commitment, right?).

In case we plan to use that column for Burn-Down Commitment Purchase as well, then we need to know how to distinguish whether a particular element in the JSON represents a commitment purchase or a contribution to a Build-Up commitment.

---

**Just as a side note:** we're currently considering modeling Marketplace Purchase as a negotiated Contract. In the [conceptual sample data](https://docs.google.com/spreadsheets/d/1PWw4K3E6BT5ZVq8CrPd3TbGq1Cs59i5wvRgmFvj03Rg/edit?gid=9833108#gid=9833108), Burn-Down Commitment Purchase information is temporarily provided via the Contract Applied column, and the ResourceId is set to carry the ContractCommitmentID, which helps us identify what was actually purchased.

Author: @ijurica Contract Applied is meant to handle for the application of any contract type.  Maybe we can meet 1:1 next week and review what you're describing to make sure we handle for your needs, and then I/we can author examples to taste?

Author: Discussed in Aug 26 TF1.  @ijurica and I will meet 1:1 and flesh this out for purchases, both the model and the examples.

Author: We met and discussed this!  I have made changes to this draft per that discussion.  Resolving this comment.  @ijurica please unresolve if you feel there is more to discuss.

The datapoints are:

* `Contract Commitment ID`: The unique identifier representing the contract line item.
* `Contract Commitment Applied Cost`: The value of the charge applied to the contract line item.
* `Contract Commitment Applied Quantity`: The usage of the charge applied to the contract line item.
* `Contract Commitment Applied Unit`: The unit of measure for the usage of the charge applied to the contract line item.

The ContractApplied column adheres to the following requirements:

* ContractApplied MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractApplied MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* ContractApplied property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* ContractApplied nullability is defined as follows:
  * ContractApplied MUST be null when [ContractID](#contractid) is null.
  * ContractApplied MUST NOT be null when ContractID is not null.
* When ContractApplied is not null, ContractApplied adheres to the following additional requirements:

>   * ContractApplied MUST contain four key-value pairs, representing ContractCommitmentID, ContractCommitmentAppliedCost, ContractCommitmentAppliedQuantity, and ContractCommitmentAppliedUnit.

Reviewer 2: We discussed that contact applied cost, quantity, and unit could be always redundant with other columns for some providers.
To mitigate this, an idea could be to introduce the notion of "calculated column" in FOCUS meta-data: a meta-data attribute for a column could specify how to calculate this column from other columns. This would save space in the FOCUS dataset size at import time from provider. In the practitioner data lake, the column could be actually populated from the calculation (use "Custom Column" in PowerBI) or calculated on-the-fly when needed (use "Measure" in PowerBI).

Author: Thank you for detailing this here, @marc-perreaut.

I see two cons with that approach:

* Increased complexity for the practitioner to determine which column is applicable, as it requires a data prep step in order to start leveraging the information.  (Not ideal, but also maybe a huge deal.)
* Does not allow for differences in _which_ cost is applied from one Contract Commitment to the next.  (This, to me, feels like a deal breaker.  If Commitment 1 uses Billed Cost and Commitment 2 uses Effective Cost, then this solution does not work.)

Ultimately, I recommend keeping the solution as I have modeled it.  To my mind, the simplicity and flexibility outweighs the need to strictly minimize dataset size.

Reviewer 2: To clarify, I did not propose to change the solution, but to introduce a a new feature in column meta-data to save bandwidth/space when the value of a column is always the same as another column. Another use case is when Contracted Cost is always the same as List Cost.

Author: Yes, that was understood!  It does not change my point: that would require the multi-provider practitioner to perform a data transformation in _some_ but not _all_ of their FOCUS Cost and Usage datasets in order to put all the data together.  To my mind, the dataset size savings do not justify that added complexity.

Reviewer 2: Ok then I would propose to resolve this comment, and maybe debate in a new issue whether a bit of logic at meta-data level is worth to save space for column values that are constant or duplicate of others for some providers.

Author: Sounds good.  @Matt-Cowsert, can you please work with @marc-perreaut to capture an issue in the backlog for this -- then we can resolve this comment?

Reviewer 3: I'll work with @marc-perreaut on crafting a new PR for this use case.

Author: Thanks!  Resolving this comment.

> ## JSON Datapoints

Author: Per @cnharris10: add carveout for provider-provided columns.  Also enforce scalars (don't allow nesting).

Author: I have addressed this.  Resolving -- though we may need to revisit.

The ContractCommitmentAppliedUnit column adheres to the following requirements:

* ContractCommitmentAppliedUnit MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractCommitmentAppliedUnit MUST be of type String.
* ContractCommitmentAppliedUnit MUST conform to [StringHandling](#stringhandling) requirements.
* ContractCommitmentAppliedUnit SHOULD conform to [UnitFormat](#unitformat) requirements.
* ContractCommitmentAppliedUnit nullability is defined as follows:
  * ContractCommitmentAppliedUnit MUST be null when ContractCommitmentAppliedQuantity is null.
  * ContractCommitmentAppliedUnit MUST NOT be null when ContractCommitmentAppliedQuantity is not null.

# Examples

A single charge has two contractual commitments applied:

1) Spend a certain amount of money on the relevant service.
2) Consume a certain number of compute hours on the relevant resource type.

```json
{
"ContractApplied": [
          {
               "ContractCommitmentID": "12346",
               "ContractCommitmentAppliedCost": 100.00,
               "ContractCommitmentAppliedQuantity": null,

>                "ContractCommitmentAppliedUnit": null

Author: Note the nulls here.  A couple questions to discuss:

 - Shall we allow null in the Quantity or Cost (requiring at least one of them to be not null)?  
 - Shall we allow null in the Unit if Quantity is null?

Author: Discussed in Aug 19 TF1.  Outcome: allow nulls in Quantity or Cost (but not both), and allow null in Unit if Quantity is null.  Will revise requirements and resolve this.

Author: I have added requirements to handle for this, but I will confer with @ijurica to confirm that I have laid it out correctly.

Author: Discussed.  Will leverage Spend and Usage in the modification of these.

           },
           {
               "ContractCommitmentID": "23456",
               "ContractCommitmentAppliedCost": null,
               "ContractCommitmentAppliedQuantity": 10,
               "ContractCommitmentAppliedUnit": "compute_hours"
           }
     ]

> }

Reviewer: **Suggestion:** We might consider using an alternative structure where ContractApplied is an object, keyed by ContractCommitmentId, and the ID is also retained as a property inside each object.

- **PROs of the alternative approach:**
  - Enables direct access by ID without iteration and improves lookup performance
  - Enforces uniqueness of ContractCommitmentId at the structure level
- **CONs:**
  - Ordering is not guaranteed (it that even matters in this case)
  - Might be slightly more complex to iterate over

```suggestion
{
    "ContractApplied": [
        "12346": {
            "ContractCommitmentId": "12346",
            "ContractCommitmentAppliedCost": 100.00,
            "ContractCommitmentAppliedQuantity": null,
            "ContractCommitmentAppliedUnit": null
        },
        "23456": {
            "ContractCommitmentId": "23456",
            "ContractCommitmentAppliedCost": null,
            "ContractCommitmentAppliedQuantity": 10,
            "ContractCommitmentAppliedUnit": "compute_hours"
        }
    }
}
```

---

**Note: **

If we wanted to avoid redundancy, we could omit the ContractCommitmentId from the object body and rely solely on the key. 

- **PROs of keeping the ID as a property inside the object:**
  - Each object self-contained, which is useful for downstream processing.
  - Simplifies transformation back to an array if needed.
  - Maintains consistency
- **CONs**
  - Redundancy and risk of inconsistency between key and internal ID

Author: Very interesting, idea, @ijurica -- thank you for sharing!

@mike-finopsorg, what do you think about this?  Does it help assuage a performance concern?  Or does it create more challenges than it solves?

Reviewer 4: This was the back and forth we had the other week with @rileyjenk, both have pros and cons. My research into this found that Array of Objects while harder to find the object of the exact known ID, is much better for unnesting using database built ins. So I sided in the end that Array of objects is better than an Object of Objects.

Author: Got it -- yes, I remember.  @ijurica: @mike-finopsorg did some research (ChatGPT output [here](https://chatgpt.com/share/68ad246c-1f78-8004-a353-6c6ff1ff35f1)) and determined that support for arrays is more widespread in various data cloud solutions -- enough so that that is the solution I've proposed here.

Author: Discussed with @ijurica.  Resolving.

```

## Column ID

ContractApplied

## Display Name

Contract Applied

## Description

A set of four datapoints that associate a charge to one or more [*contract commitments*](#glossary:contract-commitment).

## Content Constraints

| Constraint    | Value                              |
| :------------ | :--------------------------------- |
| Column type   | Dimension and Metric               |
| Feature level | Conditional                        |
| Allows nulls  | True                               |

> | Data type     | JSON                               |

Author: @kk09v has proposed implementing a `JSON Object Format` attribute per #1383.  If/when we approve that work, we would want to leverage it here.

# Contract Applied

Contract Applied is a set of four datapoints that associate a charge to one or more [*contract commitments*](#glossary:contract-commitment), denoted as key-value pairs in JSON format.  Contract Applied allows the practitioner to track the progress of the commitments to which they have agreed with a provider.

The datapoints are:

> * `Contract Commitment ID`: The unique identifier representing the contract line item.

Author: @rileyjenk's feedback from #1349 on [this comment](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1349/files#r2278178894):

> Is line item too nebulous? A line item may be interpreted as a specific discount, allocation, or feature requirement and not specifically what we would term as a commitment. Should we turn commitment into Commitment and have it be a reserved term in FOCUS land?

My response:

> That's worth discussing. It's possible that a 0 cost, 0 quantity line item that may be on a contract (think "free box of cigars every year"). It's not a commitment per se, but it's a benefit that results from the other commitments. It's not directly applicable to cost and usage, though, so I could see the argument of throwing it out. Open question in my mind right now. What do you think?

Let's put further discussion here, given that I've closed #1349 for now.

Reviewer 4: ```suggestion
* `Contract Commitment ID`: The unique identifier representing a single contract term.
```

The Contract Commitment Applied Unit represents a provider-specified measurement unit for the usage declared in Contract Commitment Applied Quantity. Contract Commitment Applied Unit complements the Contract Commitment Applied Quantity metric.

The ContractCommitmentAppliedUnit column adheres to the following requirements:

* ContractCommitmentAppliedUnit MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractCommitmentAppliedUnit MUST be of type String.
* ContractCommitmentAppliedUnit MUST conform to [StringHandling](#stringhandling) requirements.
* ContractCommitmentAppliedUnit SHOULD conform to [UnitFormat](#unitformat) requirements.
* ContractCommitmentAppliedUnit nullability is defined as follows:
  * ContractCommitmentAppliedUnit MUST be null when ContractCommitmentAppliedQuantity is null.
  * ContractCommitmentAppliedUnit MUST NOT be null when ContractCommitmentAppliedQuantity is not null.

# Examples

A single charge has two contractual commitments applied:  

1) Spend a certain amount of money on the relevant service. 
2) Consume a certain number of compute hours on the relevant resource type.

> ```json

Author: @rileyjenk @mike-finopsorg 
How does this look?  Good?  Bad?

Reviewer 4: ```
"ContractApplied": [
          {
               "ContractCommitmentID": "12346", 
               "ContractCommitmentAppliedCost": 100.00, 
               "ContractCommitmentAppliedQuantity" : 0, 
               "ContractCommitmentAppliedUnit" : "compute_hours"
           },
           {
               "ContractCommitmentID": "23456", 
               "ContractCommitmentAppliedCost": 0, 
               "ContractCommitmentAppliedQuantity" : 10, 
               "ContractCommitmentAppliedUnit" : "compute_hours"
           }
     ]
```

Author: Example applied!  Resolving.

* `Contract Commitment Applied Cost`: The value of the charge applied to the contract line item.
* `Contract Commitment Applied Quantity`: The usage of the charge applied to the contract line item.
* `Contract Commitment Applied Unit`: The unit of measure for the usage of the charge applied to the contract line item.

The ContractApplied column adheres to the following requirements:

* ContractApplied MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractApplied MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* ContractApplied property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* ContractApplied nullability is defined as follows:
  * ContractApplied MUST be null when [ContractID](#contractid) is null.
  * ContractApplied MUST NOT be null when ContractID is not null.
* When ContractApplied is not null, ContractApplied adheres to the following additional requirements:
  * ContractApplied MUST contain four key-value pairs, representing ContractCommitmentID, ContractCommitmentAppliedCost, ContractCommitmentAppliedQuantity, and ContractCommitmentAppliedUnit.

>   * ContractApplied MAY contain further key-value pairs, representing custom datapoints provided by the data generator.

Author: What normative requirements would we add here around naming convention?  @ijurica @kk09v any guidance you can provide me from the work done on `SkuPriceDetails`?

Author: Discussed in Aug 26 TF1.  Need to add guidance for `x_` prefix, similar to `SkuPriceDetails`.

Author: Added.  Resolving.

The ContractApplied column adheres to the following requirements:

* ContractApplied MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractApplied MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* ContractApplied property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* ContractApplied nullability is defined as follows:
  * ContractApplied MUST be null when [ContractID](#contractid) is null.
  * ContractApplied MUST NOT be null when ContractID is not null.
* When ContractApplied is not null, ContractApplied adheres to the following additional requirements:
  * ContractApplied MUST contain four key-value pairs, representing ContractCommitmentID, ContractCommitmentAppliedCost, ContractCommitmentAppliedQuantity, and ContractCommitmentAppliedUnit.
  * ContractApplied MAY contain further key-value pairs, representing custom datapoints provided by the data generator.
  * When ContractApplied custom key-value pairs are present:
    * ContractApplied custom key-value pairs MUST be documented by the data generator.

>     * ContractApplied custom key-value pairs MUST NOT be nested.

Author: We are going to use `JSON Object` format as proposed in #1383.  I need to revisit this and leverage that verbiage.

Contract Commitment Applied Cost represents the cost of the charge applied to the contract line item.  Contract Commitment Applied Cost is associated with the contract line item via Contract Commitment ID.  Contract Commitment Applied Cost is commonly used for monitoring the progress towards fulfilling contractual commitments that may facilitate discounts for [*resources*](#glossary:resource) or [*services*](#glossary:service) as negotiated between a provider and a customer.

The ContractCommitmentAppliedCost column adheres to the following requirements:

* ContractCommitmentAppliedCost MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractCommitmentAppliedCost MUST be of type Decimal.
* ContractCommitmentAppliedCost MUST conform to [NumericFormat](#numericformat) requirements.
* ContractCommitmentAppliedCost nullability is defined as follows:
  * ContractCommitmentAppliedCost MUST be null when ContractCommitmentID is null.
  * ContractCommitmentAppliedCost MUST NOT be null when ContractCommitmentID is not null.
  * ContractCommitmentAppliedCost MUST NOT be null when ContractCommitmentQuantity is null.

>   * ContractCommitmentAppliedCost MAY be null when ContractCommitmentQuantity is not null.

Author: @ijurica Is this a sufficient way in which to state that one of either Cost or Quantity must be not null?  Or should I present this differently?

Author: @ijurica believes we should tie this to the Category of `Spend` or `Usage`.

Author: @ijurica I am not sure how we might tie this to a Category of `Spend` or `Usage`.  If you have a suggestion, I'm happy to consider it!  In the meantime, I have elected to tie the nullability of Cost to the not nullability of Quantity, and vice versa.  Resolving this comment.

The datapoints are:

* `Contract Commitment ID`: The unique identifier representing a single contract term.
* `Contract Commitment Applied Cost`: The value of the charge applied to a single contract term.
* `Contract Commitment Applied Quantity`: The usage of the charge applied to a single contract term.
* `Contract Commitment Applied Unit`: The unit of measure for the usage of the charge applied to a single contract term.

The ContractApplied column adheres to the following requirements:

* ContractApplied MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractApplied MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* ContractApplied property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* ContractApplied nullability is defined as follows:
  * ContractApplied MUST be null when [ContractID](#contractid) is null.
  * ContractApplied MUST NOT be null when ContractID is not null.
* When ContractApplied is not null, ContractApplied adheres to the following additional requirements:

>   * ContractApplied MUST contain four key-value pairs, representing ContractCommitmentID, ContractCommitmentAppliedCost, ContractCommitmentAppliedQuantity, and ContractCommitmentAppliedUnit.
>   * ContractApplied MAY contain further key-value pairs, representing custom datapoints provided by the data generator.

Author: ```suggestion
  * The following requirements apply to the objects of the ContractApplied array:
    * ContractApplied objects MUST contain four key-value pairs, representing ContractCommitmentID, ContractCommitmentAppliedCost, ContractCommitmentAppliedQuantity, and ContractCommitmentAppliedUnit.
    * ContractApplied objects MAY contain additional key-value pairs, representing custom datapoints provided by the data generator.
```

A Contract Commitment ID is a provider-assigned identifier describing an agreement negotiated between a provider and a customer.  Contracts can include commitment to a certain amount of spend or usage over an agreed period of time.

The ContractCommitmentID column adheres to the following requirements:

* ContractCommitmentID MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractCommitmentID MUST be of type String.
* ContractCommitmentID MUST conform to [StringHandling](#stringhandling) requirements.
* ContractCommitmentID nullability is defined as follows:
  * ContractCommitmentID MUST be null when a [*charge*](#glossary:charge) is not related to a *contract commitment*.
  * ContractCommitmentID MUST NOT be null when a *charge* is related to a *contract commitment*.
* When ContractCommitmentID is not null, ContractCommitmentID adheres to the following additional requirements:
  * ContractCommitmentID MUST be a unique identifier within the provider.
  * ContractCommitmentID SHOULD be a fully-qualified identifier.
* ContractCommitmentID MUST have one and only one parent [ContractID](#contractid).

> * ContractCommitmentID MAY be equal to ContractID.

Author: Per discussion with @ijurica: do we need a requirement to put the `Contract Commitment ID` in the `Resource ID` column in the case of contract commitment purchase?

Author: Added a requirement to handle for this.  Resolving!

  * ContractCommitmentAppliedQuantity MAY be null when ContractCommitmentCost is not null.
* ContractCommitmentAppliedQuantity MUST be a valid decimal value.

### Contract Commitment Applied Unit

The Contract Commitment Applied Unit represents a provider-specified measurement unit for the usage declared in Contract Commitment Applied Quantity. Contract Commitment Applied Unit complements the Contract Commitment Applied Quantity metric.

The ContractCommitmentAppliedUnit column adheres to the following requirements:

* ContractCommitmentAppliedUnit MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractCommitmentAppliedUnit MUST be of type String.
* ContractCommitmentAppliedUnit MUST conform to [StringHandling](#stringhandling) requirements.
* ContractCommitmentAppliedUnit SHOULD conform to [UnitFormat](#unitformat) requirements.
* ContractCommitmentAppliedUnit nullability is defined as follows:
  * ContractCommitmentAppliedUnit MUST be null when ContractCommitmentAppliedQuantity is null.
  * ContractCommitmentAppliedUnit MUST NOT be null when ContractCommitmentAppliedQuantity is not null.

# Examples

> ## Example 1

Author: Need an example for a purchase.

Author: Need an example laying out how a purchase of a contract commitment could still have three objects in the Contract Applied array, where one of the objects applies to the purchase, and the other two might represent building up to other commitments.  In this case, the practitioner would need to reference something to know which entry is the purchase record, and that could be accomplished by putting the Contract Commitment ID in the Resource ID field and then that tells you which one is the purchase vs consumption/usage/buildup.  This would follow the pattern established for Commitment Discount ID.

Author: Have augmented the examples!  We may need to do more, but I'm resolving this comment.

1) Spend a certain amount of money on the relevant service.
2) Consume a certain number of compute hours on the relevant resource type.

```json
{
"ContractApplied": [
          {
               "ContractCommitmentID": "12346",
               "ContractCommitmentAppliedCost": 100.00,
               "ContractCommitmentAppliedQuantity": null,
               "ContractCommitmentAppliedUnit": null
           },
           {
               "ContractCommitmentID": "23456",
               "ContractCommitmentAppliedCost": null,
               "ContractCommitmentAppliedQuantity": 10,
               "ContractCommitmentAppliedUnit": "compute_hours"
           }
     ]
}
```

## Example 2

The same as Example 1, except a custom key-value pair `x_ContractCommitmentCostBalance` is provided by the data generator.   This datapoint represents the value remaining on a burndown contract commitment, which started at $500k.

```json
{
"ContractApplied": [
          {
               "ContractCommitmentID": "12346",
               "ContractCommitmentAppliedCost": 100.00,
               "ContractCommitmentAppliedQuantity": null,
               "ContractCommitmentAppliedUnit": null,

>                "x_ContractCommitmentCostBalance": 499000.00

Author: ```suggestion
               "x_ContractCommitmentCostBalance": 499900.00
```

           },
           {
               "ContractCommitmentID": "23456",
               "ContractCommitmentAppliedCost": null,
               "ContractCommitmentAppliedQuantity": 10,
               "ContractCommitmentAppliedUnit": "compute_hours",
               "x_ContractCommitmentCostBalance": null
           }
     ]
}
```

## Column ID

ContractApplied

## Display Name

Contract Applied

## Description

A set of four datapoints that associate a charge to one or more [*contract commitments*](#glossary:contract-commitment).

## Content Constraints

| Constraint    | Value                              |
| :------------ | :--------------------------------- |
| Column type   | Dimension and Metric               |
| Feature level | Conditional                        |
| Allows nulls  | True                               |
| Data type     | JSON                               |

> | Value format  | [KeyValueFormat](#key-valueformat) |

Author: ```suggestion
| Value format  | [JSONObjectFormat](#jsonobjectformat) |
```

# Contract Applied

> Contract Applied is a set of datapoints that associate a charge to one or more [*contract commitments*](#glossary:contract-commitment), denoted as key-value pairs in JSON format.  Contract Applied allows the practitioner to track the progress of the commitments to which they have agreed with a provider.

Author: Per @ijurica: do we need to be explicit that this is not inclusive of commitment discounts?

Author: How rigorously would we say the provider should or should not put commitment discounts here?

Author: Per @flanakin: don't say anything one way or the other, and revisit in 1.4 or a future release.

Author: My opinion: while it doesn't "break" anything to put commitment discounts here, it would introduce two versions of the truth for commitment discounts, given that we have data in two places.  But there would be a pro for having all commitments denoted in a single place.

Author: Upon reflection, I think it best to proceed as @flanakin suggests: forego a recommendation one way or the other, and revisit in a future release.

I will leave this open until Sep 9 TF1, then close.

Author: No further opinions expressed in Sep 9 TF1, other than @ijurica's agreement that this is the best path forward.  Resolving.

* ContractCommitmentAppliedUnit MUST be of type String.
* ContractCommitmentAppliedUnit MUST conform to [StringHandling](#stringhandling) requirements.
* ContractCommitmentAppliedUnit SHOULD conform to [UnitFormat](#unitformat) requirements.
* ContractCommitmentAppliedUnit nullability is defined as follows:
  * ContractCommitmentAppliedUnit MUST be null when ContractCommitmentAppliedQuantity is null.
  * ContractCommitmentAppliedUnit MUST NOT be null when ContractCommitmentAppliedQuantity is not null.

# Examples

## Example 1: Initial contract commitment

A single Cost and Usage charge represents the values stated on a contract and its three contract commitments agreed between a provider and a customer:

1) 12345: Spend $500k overall.  (This is the value of the contract, and thus ContractID = ContractCommitmentID.)
2) 23456: Spend $25k on a particular service.
3) 34567: Consume 100k compute hours on a particular resource type.

The Charge Category is denoted as Purchase, and the Contract ID, Resource ID, and Contract Commitment ID are all denoted as 12345.

```json
{
  "ResourceID": "12345",
  "ChargeCategory": "Purchase",

>   "ContractID": "12345",

Author: Per @flanakin: is there an example of multiple contracts applied to a single cost and usage row?  Splitting records would be problematic, per multiple buildup contract commitment examples.

Author: Addressed by demoting ContractID to inside the array of objects.  Resolving.

>   "ContractApplied": [

Author: Per @ijurica: include Billed and Effective cost columns in these examples.

Author: Done!  Resolving.

> # Contract Applied

Author: The organization of the the content on this page will be dependent on a reconciliation I will need to perform with @kk09v and @qquigley as they steward PR #1394, which also leverages the JSON Format as proposed in PR #1383.  That can be done during consistency review, though I do plan to complete that before approval.  The goal for now is to obtain consensus on the content itself.

Author: Captured this action in a separate comment.  Resolving this.

               "ContractCommitmentID": "23456",
               "ContractCommitmentAppliedCost": 15.00,
               "ContractCommitmentAppliedQuantity": null,
               "ContractCommitmentAppliedUnit": null,
               "x_ContractCommitmentCostBalance": 24985.00
           },
           {
               "ContractID": "12345",
               "ContractCommitmentID": "34567",
               "ContractCommitmentAppliedCost": null,
               "ContractCommitmentAppliedQuantity": 0.50,
               "ContractCommitmentAppliedUnit": "compute_hours",
               "x_ContractCommitmentCostBalance": null
           }
     ]
}
```

## Column ID

ContractApplied

## Display Name

Contract Applied

## Description

A set of datapoints that associate a charge to one or more [*contract commitments*](#glossary:contract-commitment).

## Content Constraints

| Constraint    | Value                              |
| :------------ | :--------------------------------- |
| Column type   | Dimension and Metric               |
| Feature level | Conditional                        |
| Allows nulls  | True                               |
| Data type     | JSON                               |

> | Value format  | [JSONFormat](#jsonformat)          |

Author: The leveraging of the JSON Format here is dependent on the approval of PR #1383.  This may need to be addressed during consistency review, depending on the timing of that approval.

Author: @kk09v and I will sync today to finalize the JSON Object format and then apply it here.

Author: We synced on this and resolved.  Closing!

<b>Contract ID</b>

Contract ID is a provider-assigned identifier for a contract describing the agreed terms between a provider and a customer.  Contracts can include commitment to a certain amount of spend or usage over an agreed period of time.

The "ContractId" property adheres to the following requirements:

* "ContractId" MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* "ContractId" MUST be of type String.
* "ContractId" MUST conform to [StringHandling](#stringhandling) requirements.
* "ContractId" nullability is defined as follows:
  * "ContractId" MUST be null when a [*charge*](#glossary:charge) is not related to a *contract commitment*.
  * "ContractId" MUST NOT be null when a *charge* is related to a *contract commitment*.
* When "ContractId" is not null, "ContractId" adheres to the following additional requirements:
  * "ContractId" MUST be a unique identifier within the provider.
  * "ContractId" SHOULD be a fully-qualified identifier.

<b>Contract Commitment ID</b>

> A Contract Commitment ID is a provider-assigned identifier describing an agreement negotiated between a provider and a customer.  Contracts can include commitment to a certain amount of spend or usage over an agreed period of time.

Author: Per @ijurica:
```suggestion
A Contract Commitment ID is a provider-assigned identifier describing an agreement agreed between a provider and a customer.  Contracts can include commitment to a certain amount of spend or usage over an agreed period of time.
```

The "ContractCommitmentID" property adheres to the following requirements:

* "ContractCommitmentID" MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* "ContractCommitmentID" MUST be of type String.
* "ContractCommitmentID" MUST conform to [StringHandling](#stringhandling) requirements.
* "ContractCommitmentID" nullability is defined as follows:
  * "ContractCommitmentID" MUST be null when a [*charge*](#glossary:charge) is not related to a *contract commitment*.
  * "ContractCommitmentID" MUST NOT be null when a *charge* is related to a *contract commitment*.
* When "ContractCommitmentID" is not null, "ContractCommitmentID" adheres to the following additional requirements:
  * "ContractCommitmentID" MUST be a unique identifier within the provider.
  * "ContractCommitmentID" SHOULD be a fully-qualified identifier.
  * "ContractCommitmentID" MUST have one and only one parent "ContractID".
  * "ContractCommitmentID" MUST be equal to ResourceID when ChargeCategory is "Purchase".
  * "ContractCommitmentID" MAY be equal to "ContractID".

<b>Contract Commitment Applied Cost</b>

> Contract Commitment Applied Cost represents the cost of the charge applied to the contract line item.  Contract Commitment Applied Cost is associated with the contract line item via Contract Commitment ID.  Contract Commitment Applied Cost is commonly used for monitoring the progress towards fulfilling contractual commitments that may facilitate discounts for [*resources*](#glossary:resource) or [*services*](#glossary:service) as negotiated between a provider and a customer.

Reviewer: I suggest we avoid using the term `negotiated`, at least in this iteration:

```suggestion
Contract Commitment Applied Cost represents the cost of the charge applied to the contract line item.  Contract Commitment Applied Cost is associated with the contract line item via Contract Commitment ID.  Contract Commitment Applied Cost is commonly used for monitoring the progress towards fulfilling contractual commitments that may facilitate discounts for [*resources*](#glossary:resource) or [*services*](#glossary:service) as agreed between a provider and a customer.
```

The "ContractCommitmentAppliedCost" property adheres to the following requirements:

* "ContractCommitmentAppliedCost" MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* "ContractCommitmentAppliedCost" MUST be of type Decimal.
* "ContractCommitmentAppliedCost" MUST conform to [NumericFormat](#numericformat) requirements.
* "ContractCommitmentAppliedCost" nullability is defined as follows:
  * "ContractCommitmentAppliedCost" MUST NOT be null when "ContractCommitmentAppliedQuantity" is null.
  * "ContractCommitmentAppliedCost" MAY be null in all other cases.
* "ContractCommitmentAppliedCost" MUST be a valid decimal value.
* "ContractCommitmentAppliedCost" MUST be denominated in the BillingCurrency.

<b>Contract Commitment Applied Quantity</b>

> Contract Commitment Applied Quantity represents the quantity of the charge applied to the contract line item.  Contract Commitment Applied Quantity is associated with the contract line item via Contract Commitment ID.  Contract Commitment Applied Quantity is commonly used for monitoring the progress towards fulfilling contractual commitments that may facilitate discounts for [*resources*](#glossary:resource) or [*services*](#glossary:service) as negotiated between a provider and a customer.

Reviewer: I suggest we avoid using the term `negotiated`, at least in this iteration:

```suggestion
Contract Commitment Applied Quantity represents the quantity of the charge applied to the contract line item.  Contract Commitment Applied Quantity is associated with the contract line item via Contract Commitment ID.  Contract Commitment Applied Quantity is commonly used for monitoring the progress towards fulfilling contractual commitments that may facilitate discounts for [*resources*](#glossary:resource) or [*services*](#glossary:service) as agreed between a provider and a customer.
```

# Contract Applied

> Contract Applied is a set of properties that associate a charge with one or more [*contract commitments*](#glossary:contract-commitment), denoted as key-value pairs in a JSON object.  Contract Applied allows the practitioner to track the progress of the commitments to which they have agreed with a provider.

Reviewer: `contract-commitment` glossary term might be missing

Author: Agreed... I have been deferring to your work to be completed per FR #1070 / PR #1390, which is precisely meant to augment the glossary for this purpose.  I'd prefer to get that done and then leverage it through application in consistency review.

Author: Discussed in Sep 16 TF1.  At a minimum, need entries for `Contract` and `Contract Commitment` in the glossary to make sure that the links to the glossary work.

Author: @ijurica I have now added entries for Contract and Contract Commitment.  Please review and resolve if this looks good to you!

Author: @ijurica added a suggested revision, which I have now committed.  Resolving this thread!

* `Contract ID`: The unique identifier representing a single contract.
* `Contract Commitment ID`: The unique identifier representing a single contract term.
* `Contract Commitment Applied Cost`: The value of the charge applied to a single contract term.
* `Contract Commitment Applied Quantity`: The usage of the charge applied to a single contract term.
* `Contract Commitment Applied Unit`: The unit of measure for the usage of the charge applied to a single contract term.

In addition to these, a data generator may include one or more custom properties, also denoted as key-value pairs.

## Requirements

### Column Requirements

The ContractApplied column adheres to the following requirements:

* ContractApplied MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports *contract commitments*.
* ContractApplied MUST conform to [KeyValueFormat](#key-valueformat) requirements.
* ContractApplied property keys SHOULD conform to [PascalCase](#glossary:pascalcase) format.
* ContractApplied MUST NOT be null when a contract is applied to the charge.
* When ContractApplied is not null, ContractApplied adheres to the following additional requirements:

>   * ContractApplied objects MUST contain four key-value pairs, representing ContractCommitmentID, ContractCommitmentAppliedCost, ContractCommitmentAppliedQuantity, and ContractCommitmentAppliedUnit.

Author: Per #1462, I need to add double quotes around these.

  * ContractApplied objects MAY contain custom key-value pairs, representing additional datapoints provided by the data generator.
  * When ContractApplied custom key-value pairs are present:
    * ContractApplied custom key-value pairs MUST be prefixed with a consistent `x_` prefix to identify them as external, custom columns and distinguish them from FOCUS columns to avoid conflicts in future releases.
    * ContractApplied custom key-value pairs MUST be documented by the data generator.
    * ContractApplied custom key-value pairs MUST NOT be nested.

### Object Schema Requirements

> Allocated Method Details consists of a valid JSON object which contains an array of key-value objects describing the one or more contract commitments applied to the charge. Each object consists of FOCUS-defined keys but can be extended to provide additional details about the contract application.

Reviewer: ```suggestion
Contract Applied consists of a valid JSON object which contains an array of key-value objects describing the one or more contract commitments applied to the charge. Each object consists of FOCUS-defined keys but can be extended to provide additional details about the contract application.
```

* If ContractApplied is not null, ContractApplied adheres to the following requirements:
  * ContractApplied MUST have a top-level key "Elements" which contains an array.
  * Each item in "Elements" MUST be an object.
    * Objects inside "Elements" MUST conform to [KeyValueFormat](#key-valueformat) requirements.

>     * Objects inside "Elements" MUST contain key-value pairs (contract applications).

Reviewer: The current wording suggests that each key-value pair is a contract application. However, the requirements that follow indicate that each Element represents a contract application, while the key-value pairs are actually **contract application properties**.

```suggestion
    * Objects inside "Elements" MUST contain key-value pairs (contract application properties).
```

      * FOCUS-defined contract application properties adhere to the following additional requirements:
        * Contract application property key MUST match the spelling and casing specified for the FOCUS-defined property.
        * Contract application property value MUST be of the type specified for that property.
        * Contract application property MUST adhere to additional normative requirements specific to that property.
      * Contract application property keys MUST begin with the string "x_" unless it is a FOCUS-defined allocation property.
  * ContractApplied root object MAY contain custom properties, in addition to "Elements".

### Content Requirements

The following keys are used for contract application properties to facilitate querying data across allocations and across providers. FOCUS-defined keys will appear in the list below, and custom keys will be prefixed with "x_" to make them easy to identify as well as prevent collisions.

> | Datapoint                            | Column Type | Feature Level | Allows Nulls | Data Type |

Author: Changing from `Datapoint` to `Key`, per @kk09v's suggestion.  Also, changing `Column Type` to `Key Type`.
```suggestion
| Key                                  | Key Type    | Feature Level | Allows Nulls | Data Type |
```

> | Datapoint                            | Column Type | Feature Level | Allows Nulls | Data Type |
> | ------------------------------------ | ----------- | ------------- | ------------ | --------- |
> | Contract ID                          | Dimension   | Conditional   | False        | String    |
> | Contract Commitment ID               | Dimension   | Conditional   | False        | String    |
> | Contract Commitment Applied Cost     | Dimension   | Conditional   | True         | Numeric   |
> | Contract Commitment Applied Quantity | Dimension   | Conditional   | True         | Numeric   |
> | Contract Commitment Applied Unit     | Dimension   | Conditional   | True         | String    |
> 

Author: Striking this section in deference to the Overview section farther down, per @kk09v's suggestion.

```suggestion
```

* "ContractCommitmentAppliedUnit" MUST be of type String.
* "ContractCommitmentAppliedUnit" MUST conform to [StringHandling](#stringhandling) requirements.
* "ContractCommitmentAppliedUnit" SHOULD conform to [UnitFormat](#unitformat) requirements.
* "ContractCommitmentAppliedUnit" nullability is defined as follows:
  * "ContractCommitmentAppliedUnit" MUST be null when "ContractCommitmentAppliedQuantity" is null.
  * "ContractCommitmentAppliedUnit" MUST NOT be null when "ContractCommitmentAppliedQuantity" is not null.

## Overview

### Array of Objects

The parent array is called `Elements` and contains one or more objects which communicate information about how an allocated record was calculated.

| Key | ValueType | Required | Description |
| ----- | ---- | ---------- | ----------- |
| Elements | Array | True | The parent array containing one or more objects which communicate information about how contract commitments were applied to the charge. |

### Object Entries

The `Elements` array contains one or more objects, each of which contains the following entries:

> | Key                                  | Column Type | Feature Level | Allows Nulls | Data Type |

Author: Changing from `Datapoint` to `Key`, per @kk09v's suggestion.  Also, changing `Column Type` to `Key Type`.
```suggestion
| Key                                  | Key Type    | Feature Level | Allows Nulls | Data Type |
```
